elm-format-lib/src/Parse/ParsecAdapter.hs

```diff
@@ -368,48 +345,51 @@ sourceColumn (SourcePos _ col) =
 -- Text.Parsec.Error
 
 
-data ParseError = ParseError Row Col [Message]
+data ParseError
+  = Nil
+  | Cons Message Row Col ParseError
+  | OneOf ParseError ParseError
   deriving Show
 
 
-errorPos :: ParseError -> SourcePos
-errorPos (ParseError row col _) = newPos row col
-
+parseError :: Message -> Row -> Col -> ParseError
+parseError message row col =
+  Cons message row col Nil
 
-errorMessages :: ParseError -> [Message]
-errorMessages (ParseError _ _ messages) = messages
 
+addMessage :: Message -> (Row -> Col -> ParseError) -> Row -> Col -> ParseError
+addMessage message toErr row col =
+  Cons message row col (toErr row col)
 
 
--- Create parse errors
+mergeError :: ParseError -> ParseError -> ParseError
+mergeError = OneOf
 
 
-newErrorMessage :: Message -> Row -> Col -> ParseError
-newErrorMessage msg row col
-    = ParseError row col [msg]
+errorPos :: ParseError -> SourcePos
+errorPos err =
+  case err of
+    Nil ->
+      error "An unexpeced error occured, this is likely a bug. Please report this issue at https://github.com/avh4/elm-format/issues"
```

Comment:
`errorPos` crashes on empty error. I realize that it's possible to define `ParseError` to make impossible state impossible. Since Nil isn't exposed outside this module and since it's never used to construct an empty error I figured this was ok. That said, just let me know if you'd like this changed!